### PR TITLE
Feature/bb xyz

### PR DESCRIPTION
**Pull request by jordaniza**
A few changes inbound, this will require a database update to add a `customerkey` to the DB (@Behold3th & co what do you think)

# Overview of the changes

## Fetching with customer key

We needed to introduce a way to filter our bounties based on the current url. There were a few options here:
- Use the customer Id
- Use the _id
- Use the customer name
- Use another field

Id and _id obviously work but have the disadvantage of making our urls less readable. @tescher asked if we can use the name, or a lowercased version. It's a good question, and we can still make the change relatively easily. 

**Arguments for using a dedicated customer key:**
- Customer key can be mandated to be lower case and without spaces
- Customer key can be mandated to be unique
- Customer key can be mandated not to change, or change less frequently (for historical urls)
- Does not require remembering to transform name each time

**Arguments for using a name:**
- No change in data model needed
- Names unlikely to change often, and if they do, client likely wants change in key
- A set transformation of the name can be the key (lowercase without spaces)
- Really old links are not a huge problem

## Rerouting and forwarding

General idea is that we don't build new routes, we just construct skeleton pages and import the existing components, leveraging next router. The build compilation should make this fairly lightweight.

Our general structure is seen in `packages/react-app/src/pages/dao`:

```
--------- Existing routes ----------
/ <--- bounties
/:bountyId <--- single bounty
/:bountyId/create <--- new bounty
/:bountyId/preview <--- preview before create
/:bountyId/edit

--------- New routes ----------
/dao/:customerKey/bounties/*  <---- all existing routes above are mapped with the above prefix
```

This part was relatively easy, what took time was grabbing the relevant customer data depending on all the route permutations, and making is work 'as expected'.

## Dynamic Images

Our app has so far been primarily one level deep in terms of nested routing, meaning that images could be loaded from: `./image_name.format`

This doesn't work if you are in a page like `/more/deeply/nested/page`, so we used the work done by Lemongrass to replace images dynamically:

Next allows us to fetch images from the 'public' folder provided we:
- Wrap them in a NextImage component
- Apply the proper prefixing

This change can be seen in `packages/react-app/src/components/global/Header/Logo.tsx`

## Cases to consider

Workflows (Thinking separate PRs):
* Create bounty and subsequent redirects
* Edit Bounty and subsequent redirects
* Preview ...
* Delete  ...

Unhappy paths
* DAO not found

### CODE REVIEW

File: packages/react-app/src/constants/Bankless.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import { CustomerProps } from '../models/Customer';
  */
 export const BANKLESS: CustomerProps = {
 	customerName: 'BanklessDAO',
+	customerKey: 'banklessdao',
```
Comment (tescher) [inline packages/react-app/src/constants/Bankless.ts:7]: Is there a reason for a separate key rather than just the name (or derivation on the name e.g. <name>.toLowerCase)? Do we expect collisions on customer name?

**Review by tescher — COMMENTED** [review]
Just one question so far

File: packages/react-app/src/constants/Bankless.ts (hunk 1 above)
Comment (jordaniza) [reply packages/react-app/src/constants/Bankless.ts:7]: This is a good questions and will outline thoughts above to discuss

**Comment by tescher** [comment]
Is it the intention that a user needs to be logged into Discord and a member of a server, and in the context of that particular DAO before they can create/edit/claim bounties from that DAO?

**Comment by jordaniza** [comment]
> Is it the intention that a user needs to be logged into Discord and a member of a server, and in the context of that particular DAO before they can create/edit/claim bounties from that DAO?

I think so - we seem to be going full into discord as our means of identification. 

If the DAO disables role gating, maybe it can be fully permissionless, but otherwise roles can only be applied with logins

**Comment by tescher** [comment]
> > Is it the intention that a user needs to be logged into Discord and a member of a server, and in the context of that particular DAO before they can create/edit/claim bounties from that DAO?
> 
> I think so - we seem to be going full into discord as our means of identification.
> 
> If the DAO disables role gating, maybe it can be fully permissionless, but otherwise roles can only be applied with logins

If that's the case, why do we need the customer key in the URL at all? Can't you restrict bounty access based on current session/context? A bounty is looked up with the URL, and if someone's not logged in, or isn't a member of that server, put up the appropriate message.

**Comment by jordaniza** [comment]
> > > Is it the intention that a user needs to be logged into Discord and a member of a server, and in the context of that particular DAO before they can create/edit/claim bounties from that DAO?
> > 
> > 
> > I think so - we seem to be going full into discord as our means of identification.
> > If the DAO disables role gating, maybe it can be fully permissionless, but otherwise roles can only be applied with logins
> 
> If that's the case, why do we need the customer key in the URL at all? Can't you restrict bounty access based on current session/context? A bounty is looked up with the URL, and if someone's not logged in, or isn't a member of that server, put up the appropriate message.
> 
> And if they are creating a bounty, it is created in the context of the DAO selected. Not sure what the customer in the URL is getting you.

You could do that, but it would be a marked change versus what we had before. 

As of now, all boards are public. Meaning: if you have a bounty id, you can see the bounty.

This means that people not currently logged in to discord can view the board and perform entry level interactions (get bounties, view them in discord).

I know that most SaaS applications build a clear permissions hierarchy that defaults to making different tenant data private unless otherwise specified. 

We have, until now, **not** been following that model. 

Our API is public, and we only recently introduced Role Based Access Controls now that we have state mutations supported in the app.

If we mandated that a user must be signed in on discord to access a DAO's board, that's fine, but it is something I'd want to agree with the team.

Within the scope of this piece of work, I'd suggest keeping a customer reference of some sort in the URL as an index. This allows people to view the DAO's board permissionlessly, then opt to sign in if they plan to claim. I also think that aligns with the general goal of the board, to connect DAOs to talent, both internal and external.

**Comment by tescher** [comment]
OK, I was initially missing what the purpose of the API was. If it is intended to be publicly available, then going full RESTful like you have makes sense. 

My only quibble is my mental model has customer as a property of a bounty, rather than bounties as members of a customer, so I would've treated getting a customer-filtered list of bounties like this:

bounties?customer=<customer>

instead of 
dao/<customer>/bounties

But that's just a preference thing (unless it makes auth more difficult somehow)

**Comment by jordaniza** [comment]
> OK, I was initially missing what the purpose of the API was. If it is intended to be publicly available, then going full RESTful like you have makes sense.
> 
> My only quibble is my mental model has customer as a property of a bounty, rather than bounties as members of a customer, so I would've treated getting a customer-filtered list of bounties like this:
> 
> bounties?customer=<customer>
> 
> instead of dao/<customer>/bounties
> 
> But that's just a preference thing (unless it makes auth more difficult somehow)

I really do welcome these comments as I think it's very valid. 

In practice, a lot of the API calls work exactly as you describe. But thinking from a customer's perspective I think that makes sense as it gives them more of a sense of ownership?

Seems trivial but /myName/myStuff feels, to me at least, more significant than /allStuff?name=myName

I am not wedded to the above solution, so as we work through, more than happy to keep adapting.

**Comment by tescher** [comment]
No, I think what you have is fine. Like I said, it's a preference thing unless it makes a difference under the covers somehow. I'm guessing most people have a mental model like you've implemented it. My mental model is probably the outlier.

**Comment by Behold3th** [comment]
@tescher could you give more context on why you see the relationship that
way?

I think Jordaniza and I have been subscribed to bounties belonging to a
customer since we first switched to a multi-tenant model back in Nov/Dec of
last year.

Always a good exercise to challenge our assumptions!

**Comment by tescher** [comment]
I think it's an artifact of being deep in the code on the bot. We don't deal with customers much there other than as a filter. Customer is assumed because that's the server we're on at the time. The core unit is the bounty. Almost everything is driven by the bounty.

The only reason I can see to reconsider your model in the API is either if 1) There's an auth or security reason, or 2) There's a user flow model that supports it. I don't know enough to say if \#1 is the case. For \#2, are users going to see the Bounty Board as a standalone thing where they can find bounties, and they don't necessarily have an affinity for any particular DAO, or are they going to mostly think of themselves as searching within a DAO like they do with the bot.

In the end I don't think it makes much difference since the URL structure is going to be mostly invisible except if you are trying to program against the API. Most users aren't going to know or care.

**Comment by jordaniza** [comment]
@tescher @Behold3th I was considering removing the customer key as a redundancy, a transformation on `Customer Name` would have worked equally well. 

However when researching ways to do that on Mongo, I ran into answers like this:

https://stackoverflow.com/questions/8100838/transform-mongodb-data-on-find

tl;dr - Mongo excels at reads, not data processing (This is not a SQL DB). 
If we were to use customer name, then lowercase and remove spaces as the index, we would either need to:

1. Run a mongo transformation as part of the `.find` query
2. Pull all data, for all customers, then apply the filter in memory on the server with a javascript array method.

(2) strikes me as a bad idea, as we cannot leverage anything like DB caching without rolling it ourselves.
(1) appears to be an anti-pattern from the mongo community

From what I can see, the recommended way forward is to add a new field (customerKey) in the DB to leverage fast reads. I suggest we go ahead with that data model change, then move this branch into production.

**Comment by tescher** [comment]
> From what I can see, the recommended way forward is to add a new field (customerKey) in the DB to leverage fast reads. I suggest we go ahead with that data model change, then move this branch into production.

Makes sense, although I doubt that our customer collection is ever going to be large enough to make it noticeable

**Comment by jordaniza** [comment]
> > From what I can see, the recommended way forward is to add a new field (customerKey) in the DB to leverage fast reads. I suggest we go ahead with that data model change, then move this branch into production.
> 
> Makes sense, although I doubt that our customer collection is ever going to be large enough to make it noticeable

That's the spirit :smiley:

**Comment by Behold3th** [comment]
Is the idea that you want this customerKey field to be used as the customer
key for routing in the new XYZ model? I.e ‘…/c/[…customerKey]’?

I agree that’s a better UX if customerKey is a lowercase no white space
transformation of a customer name.

The bot relies on the customerId being the discord server id so that it can
correctly route and fetch messages and users. So we will have to store a
mapping of customerKey to our current customerIds for either the bot if we
make this change or for the web if we don’t make this change.
